### PR TITLE
feat(evaluations): compare runs across datasets by shared input

### LINKED ISSUE #1946: Allow run comparison across datasets on shared cases
## Summary

A run comparison aligns cases by an intersection of case ids across the selected runs (`compare-runs-view.tsx:234-239`), so two runs that share cases are comparable regardless of which dataset they came from. Today the UI refuses that outright the moment the selected runs span more than one dataset — which is more restrictive than the alignment model needs. Worse, even if the refusal were lifted, comparison would still come up empty across datasets, because the case id itself folds in the dataset key, so the *same input* in two datasets is never the *same id*. Enabling cross-dataset comparison therefore takes two changes, not one.

## How to replicate

1. Run an evaluation on dataset A and another on dataset B, where A and B contain some identical case inputs.
2. Select both runs and open Compare (`evaluations-view.tsx:233-252`).
3. Observe the empty state (`compare-runs-view.tsx:391`): *"These runs are on different datasets (…). A run comparison needs the same dataset."* — even though the two runs share inputs that could be paired.

## Root cause

Two independent blockers:

1. **Explicit refusal.** `compare-runs-view.tsx:391` short-circuits to an empty state whenever `datasetNames.length > 1`.
2. **Case ids are dataset-scoped.** The stable case id is `tc_ + sha256(datasetKey ␀ inputCanonical ␀ occurrence)` (`traceroot-*/…/eval/ids.ts::stableCaseId`) — it **includes the dataset key**. So identical inputs in datasets A and B get different `tc_` ids, and the id-intersection (`:234-239`) is empty for genuinely cross-dataset runs even with the refusal removed.

## Impact

- Two runs that measure the same inputs under different datasets can't be compared, even though the data supports it.
- The refusal hides a deeper identity constraint (dataset-scoped ids), so "just remove the guard" would produce silently-empty comparisons.

## Direction / open questions

- Drop the hard refusal; keep an informational label when runs span datasets.
- Align cross-dataset runs on **input content** rather than the dataset-scoped `tc_` — a dataset-independent canonical-input key computed at compare time, or a content-only case identity carried alongside `tc_` (larger change — see the case-id consistency issue). Same-dataset comparison keeps using `tc_`.
- Decide how to present cases present in only one run (unpaired rows vs drop) and per-scorer deltas when the two datasets declare different scorers.

## Acceptance criteria

- Selecting runs across datasets that share inputs produces a comparison aligned on those shared inputs, instead of the "needs the same dataset" empty state.
- Same-dataset comparison behavior is unchanged.

### PR DESCRIPTION
**Pull request by dark-sorceror**
## Summary

Closes: #1946

The N-run "Run Comparison" page previously refused to compare runs that came from different datasets, and even without that refusal it couldn't have aligned them: dataset case ids (`tc_`) fold in the dataset key, so identical inputs in two datasets get different ids. This PR lets you compare runs across datasets when they share inputs — the refusal is gone, and cross-dataset runs align on the case's canonical input instead of the dataset-scoped id. Same-dataset comparison is unchanged.

## How it works

```
Selected runs → detect whether they span >1 dataset
  ├─ same dataset  → align cases by testCaseId  (unchanged common path)
  └─ cross-dataset → align cases by canonicalInputKey(input)   ← dataset-independent
        → intersection of keys present in every run (deduped, since input keys
          aren't unique within a run) → per-scorer columns, deltas, aggregates
        → non-blocking banner: "aligned by shared input, not by dataset row"
```

The alignment key is chosen once (`keyOf(r)`); everything downstream — per-scorer columns, deltas, baseline ordering, search, the result-cap partial-data banner — operates on the abstract key and is untouched. The N-run alignment lives entirely in the view; the 2-run candidate-vs-baseline engine (`comparison.ts`/`comparison-db.ts`) is not involved and was not changed.

### Frontend
- `frontend/ui/src/features/evaluations/views/compare-runs-view.tsx` — remove the different-dataset refusal; add `keyOf(r)` (testCaseId same-dataset, `canonicalInputKey` cross-dataset); dedup the intersection; add the informational banner and a cross-dataset-aware empty state
- `frontend/ui/src/lib/eval/json-value.ts` — extract the single `canonicalJson` here (a pure module a client view can import) and add `canonicalInputKey(input) = canonicalJson(decodeJsonValue(input))`
- `frontend/ui/src/lib/eval/versions.ts` — import `canonicalJson` from `json-value.ts` (one canonicalizer, behavior unchanged)

### Tests
- `frontend/ui/src/features/evaluations/compare-runs.smoke.test.tsx` — the old "refuses cross-dataset" test replaced by three cases

## Test plan

- [x] Two runs on different datasets with disjoint `testCaseId`s but shared inputs → compare (no refusal), rows align on the shared inputs, deltas compute
- [x] Cross-dataset runs that share no inputs → shared-input empty state (not the old "needs the same dataset" block)
- [x] Same-dataset runs with identical ids but edited inputs → still align by `testCaseId`, proving the common path did not switch to input-keying
- [x] `comparison.test.ts`, `comparison-db.test.ts`, `json-value.test.ts`, `versions.test.ts`, `datasets-versioning.test.ts` unaffected and green
- [x] `tsc --noEmit` clean on changed files; prettier clean

## Note

Deliberately pairs with the content-addressed-case-ids change: `tc_` stays dataset-scoped, and cross-dataset alignment intentionally keys on canonical input rather than the id. Base branch `main`.

---
## Summary by cubic
Closes #1946. Run Comparison now accepts cross-dataset runs and aligns them by shared canonical input instead of refusing them. Alignment is chosen per run against the baseline’s stable `datasetId`: baseline-dataset runs use `testCaseId`, other runs use canonical input, and the oldest selected run is the baseline.

**Comparison behavior**

- Canonical input keys normalize plaintext and JSON-encoded values plus object key order while keeping scalar types distinct.
- Duplicate baseline inputs remain separate rows; a matching cross-dataset result fills each row, while duplicate cross-run inputs keep the earliest case.
- Dataset renames do not change alignment, and same-dataset rows retain their identity in mixed selections.
- Selections with no shared inputs show a targeted empty state, and the old cross-dataset banner is removed.
- Reuses `canonicalJson` for input alignment and scorer-manifest comparison.

<sup>Written for commit 425ae44946ab14a1771bff42f5a89ac9717813e5. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/traceroot-ai/traceroot/pull/1952?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by XinweiHe** [comment]
@cubic review and rate please

File: frontend/ui/src/features/evaluations/views/compare-runs-view.tsx [hunk 1]
```diff
@@ -203,40 +206,62 @@ export function CompareRunsView({
...
-      .filter((d): d is NonNullable<typeof d> => !!d)
-      .map((d) => ({
-        run: d.run,
-        results: d.results,
-        byCase: new Map(d.results.map((r) => [r.testCaseId, r])),
-      }));
+    const bundles = loaded.map((b) => ({
+      run: b.run,
+      results: b.results,
+      byCase: new Map(b.results.map((r) => [keyOf(r), r])),
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply frontend/ui/src/features/evaluations/views/compare-runs-view.tsx]: Resolved in 7758d039d

Replaced the last-write-wins `new Map(results.map(...))` with an explicit keep-first policy: a `keepFirst` reducer builds each run's lookup map and ignores a key already present, so a run with two cases sharing a canonical input deterministically keeps the earliest occurrence instead of depending on result ordering. This was folded into the larger baseline-anchored alignment rework (row identity is now the baseline row's stable id), and a direct duplicate-key test covers the policy.

File: frontend/ui/src/features/evaluations/views/compare-runs-view.tsx [hunk 2]
```diff
@@ -203,39 +231,81 @@ export function CompareRunsView({
...
-        run: d.run,
-        results: d.results,
-        byCase: new Map(d.results.map((r) => [r.testCaseId, r])),
-      }));
-    return bundles.sort((a, b) => {
-      if (a.run.id === baselineId) return -1;
-      if (b.run.id === baselineId) return 1;
-      return a.run.runNumber - b.run.runNumber;
+    if (!baselineBundle) return [];
+    const baselineRows = [...baselineBundle.byTestCase.values()];
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply frontend/ui/src/features/evaluations/views/compare-runs-view.tsx:298]: Resolved in 7a1040dde

In cross-dataset mode alignment keys on canonical input, but the baseline row set was always `[...baselineBundle.byTestCase.values()]` — so a baseline run holding two dataset rows with the same input rendered that shared input once per row and double-counted it in the totals. The baseline rows now come from the canonical-input-keyed, keep-first `byInput` map when the comparison spans datasets (reusing the existing `crossDataset` descriptor), and stay on `byTestCase` otherwise, so same-dataset comparisons are unchanged. Added a test asserting a baseline with duplicate inputs renders one row and counts once (Duration total 2.1s, not 4.2s).

File: frontend/ui/src/features/evaluations/compare-runs.smoke.test.tsx [hunk 3]
```diff
@@ -182,18 +186,242 @@ describe("CompareRunsView — N-run diff table", () => {
...
+    );
+    mount();
+    // Cross-dataset banner confirms input-keyed alignment is in effect.
+    expect(screen.getByText(/aligned by shared input/)).toBeTruthy();
+    // The shared input renders exactly ONE row (deduped by canonical input), not two.
+    expect(screen.getAllByText(DUP)).toHaveLength(1);
+    // Counted once in the totals: baseline Duration total is 2.1s (one row @2100ms), not
+    // 4.2s (two rows). The double-counted mean would also surface as a 50.0% aggregate.
+    expect(screen.queryByText("4.2s")).toBeNull();
+    expect(screen.queryByText("50.0%")).toBeNull();
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply frontend/ui/src/features/evaluations/compare-runs.smoke.test.tsx]: Resolved in 940914e66

Good catch — the dedup test proved the shared input renders once and is counted once, but a last-write-wins map would have passed those same assertions. The fixture distinguishes the duplicates by score (dup-a routing_accuracy 100%, dup-b 0%), so I added an assertion that dup-b's 0% never surfaces (the baseline keeps dup-a's earliest 100%). I verified it fails under a last-write-wins map (renders 0.0%) and passes under the keep-first policy, so it now protects that contract directly.

File: frontend/ui/src/lib/eval/json-value.ts [hunk 4]
```diff
@@ -12,6 +12,42 @@
...
+ * two forms yields a string vs a number/boolean, so a raw `canonicalJson(decode(...))`
+ * would key them apart and fail to align. To make a legacy scalar and its JSON-encoded
+ * form converge, every SCALAR is normalized to its string content (`123` and `"123"`
+ * both → `123`); objects/arrays keep their structural, key-order-independent canonical
+ * form. The `s:` / `o:` prefixes keep a genuine string that merely looks like JSON
+ * (e.g. `"{\"a\":1}"`) from colliding with the object it resembles.
+ */
+export function canonicalInputKey(input: string): string {
+  const decoded = decodeJsonValue(input);
+  if (decoded === null || typeof decoded !== "object") return `s:${String(decoded)}`;
```
[the remark this answers is not in the document]
Comment (dark-sorceror) [reply frontend/ui/src/lib/eval/json-value.ts]: Resolved in dfc399d8e

canonicalInputKey now keys scalars by their JSON representation (`s:${JSON.stringify(decoded)}`) instead of their bare string content, so a string keeps its quotes and a JSON string, number, and boolean with the same text stay distinct (`"123"` → `s:"123"`, `123` → `s:123`; `"true"` vs `true` likewise). The legacy-plaintext convergence this replaced still holds for same-typed values — a plain-text string still matches a JSON string, and a plain-text number matches a JSON number — only the cross-type collisions (`"123"`↔`123`, `"true"`↔`true`) are dropped, which is the misalignment you flagged. Updated the unit tests to assert same-type convergence and different-type separation.

**Comment by dark-sorceror** [comment]
image


**Cross dataset comparison just by matching input content**
image

**Same dataset control**
image

**Duplicates not converging edge case**
image

Duplicate test case run identification is deferred to follow up in a comparison page revamp
